### PR TITLE
Fix corner starts

### PR DESCRIPTION
**Pull request by TomMelt**
This PR fixes corner neighbour metadata for periodic boundaries

## Key Changes

- Fixed `haloCornerBufferPositions()` logic to correctly handle periodic boundary conditions
- Updated corner neighbour detection to properly identify neighbours across periodic boundaries
- Added comprehensive tests for halo buffer positions and corner neighbour detection
- Updated reference test data for integration tests
- Improved documentation with better comments and docstrings (but more to come in #80)

## Tests

* `test_haloBufferPositions.cpp`
* `test_haloCornerBufferPositions.cpp`
* `test_is_corner_neighbour.cpp`
* `test_is_neighbour.cpp`

TODO:

- [x] add unit tests for `haloCornerBufferPositions`
- [x] add unit tests for `haloBufferPositions` functions @niravshah241 is working on
- [x] rename `haloBufferPositions` to `haloEdgeBufferPositions` (this will be handled in a following PR -- see issue #80 )
- [x] fix `haloCornerBufferPositions` for periodic corner neighbours (@niravshah241 is double checking) 
- [x] @niravshah241 is investigating why the `test_halo_corner_start` test is failing -> See https://github.com/
nextsimhub/domain_decomp/pull/78/commits/b8ef5f59ec36814b6029fe174bc70a37ba888581
- [x] @niravshah241 to resolve merge conflict
- [x] Add test diagrams in repository for easier reference later @niravshah241 
- [x] check all edge metadata in `haloBufferPositions` (i.e., use same approach as `haloCornerBufferPositions`)

### CODE REVIEW

File: Partitioner.cpp [hunk 1]
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
...
+        // or left edge. We need to check so we know where to look in the send buffer.
+        if (d1.p2.y > d2.p1.y) {
+            // this first case is true if the corner neighbour lies on the left edge
+            int dy = d1.p2.y - d2.p1.y;
+            send_pos = 2 * d2.get_width() + d2.get_height() + dy;
         } else {
-            start = left - d2.p1.x - 1;
+            // this second case works if the corner neighbour lies on the bottom edge (or in the
+            // corner of both e.g., the bottom left corner)
+            int dx = d1.p2.y - d2.p1.y;
```
Comment (niravshah241) [inline Partitioner.cpp]: Should this be `dx = d1.p2.x - d2.p1.x`?

File: Partitioner.cpp [hunk 2]
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
...
-    } else if (vertex == TOP_RIGHT) {
-        if (d2.p1.x < right) {
-            start = right - d2.p1.x;
+    } else if (vertex == TOP_LEFT) {
+        if (d1.p2.y > d2.p1.y) {
+            int dy = d1.p2.y - d2.p1.y;
+            send_pos = d2.get_width() + dy;
         } else {
-            start = (top - d2.p1.y) * d2.get_width();
+            int dx = d1.p2.y - d2.p1.y;
```
Comment (niravshah241) [inline Partitioner.cpp]: Should this be `dx = d2.p1.x - d1.p1.x - 1`?

File: Partitioner.cpp [hunk 3]
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
...
-    } else if (vertex == BOTTOM_RIGHT) {
-        if (d2.p1.y < bottom) {
-            start = (bottom - d2.p1.y - 1) * d2.get_width() + (right - d2.p1.x);
+    } else if (vertex == BOTTOM_LEFT) {
+        if (d2.p2.y > d1.p1.y) {
+            int dy = d1.p1.y - d2.p1.y;
+            send_pos = d2.get_width() + dy - 1;
         } else {
-            start = (bottom - d2.p1.y - 1) * d2.get_width();
+            int dx = d1.p2.y - d2.p1.y;
```
Comment (niravshah241) [inline Partitioner.cpp]: Should this be `dx=d1.p1.x-d2.p1.x`?

File: Partitioner.cpp [hunk 4]
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
...
-    } else if (vertex == BOTTOM_LEFT) {
-        if (d2.p1.y < bottom) {
-            start = (bottom - d2.p1.y - 1) * d2.get_width() + (left - d2.p1.x) - 1;
+    } else if (vertex == BOTTOM_RIGHT) {
+        if (d2.p2.y > d1.p1.y) {
+            int dy = d1.p1.y - d2.p1.y;
+            send_pos = 2 * d2.get_width() + d2.get_height() + dy - 1;
         } else {
-            start = (bottom - d2.p1.y) * d2.get_width() - 1;
+            int dx = d1.p2.y - d2.p1.y;
```
Comment (niravshah241) [inline Partitioner.cpp]: Should this be `dx = d1.p2.x - d2.p1.x` ?

File: Partitioner.hpp
Comment (niravshah241) [inline Partitioner.hpp]: Do we necessarily need `private` here? If no, it would be better (at least for the unit tests) if the functions `is_neighbour` , `is_corner_neighbour`, `haloBufferPositions` and `haloCornerBufferPositions` are made public.

**Review by niravshah241 — COMMENTED** [review]
Overall, looks good. Some quick formulae checks and making `private` members to `public` in `partition.hpp` have been suggested.

The unit tests for `is_corner_neighbour` and `haloCornerBufferPositions` (previously `halo_corner_start`) have been written in the branch `add-periodic-corner-nbrs`. (It will need some changes depending upon final function calls and index changes.)

File: Partitioner.cpp (hunk 1 above)
Comment (TomMelt) [reply Partitioner.cpp]: good point. I have either named `dx` incorrectly, or it should be `d1.p2.x`. I will double check these and get back to you

File: Partitioner.cpp (hunk 2 above)
Comment (TomMelt) [reply Partitioner.cpp]: same as prev comment.

File: Partitioner.cpp (hunk 3 above)
Comment (TomMelt) [reply Partitioner.cpp]: same as prev comment.

File: Partitioner.cpp (hunk 4 above)
Comment (TomMelt) [reply Partitioner.cpp]: same as prev comment.

**Comment by niravshah241** [comment]
Tests are added to this branch. The tests in `test_halo_corner_start` nonperiodic cases are fixed now https://github.com/nextsimhub/domain_decomp/pull/78/commits/b8ef5f59ec36814b6029fe174bc70a37ba888581 .

File: Partitioner.hpp
Comment (niravshah241) [reply Partitioner.hpp]: Made ```public``` from ```private```. See https://github.com/nextsimhub/domain_decomp/commit/b8ef5f59ec36814b6029fe174bc70a37ba888581 .

Comment (TomMelt) [reply Partitioner.hpp]: Thanks @niravshah241

**Comment by niravshah241** [comment]
Tests related to `is_neighbour` and `haloBufferPositions` added. There is difference in how the neighbours are considered in two functions. Given 2 subdomains `d1` and `d2`, the `RIGHT` in `is_neighbour` means `d2` is `RIGHT` of `d1` but in `haloBufferPositions`, `d1` is considered `RIGHT` of `d2`. Also, for `LEFT`, `BOTTOM` and `TOP` .

In addition the images for tests have now been added in `img` folder.

**Comment by TomMelt** [comment]
> Tests related to `is_neighbour` and `haloBufferPositions` added. There is difference in how the neighbours are considered in two functions. Given 2 subdomains `d1` and `d2`, the `RIGHT` in `is_neighbour` means `d2` is `RIGHT` of `d1` but in `haloBufferPositions`, `d1` is considered `RIGHT` of `d2`. Also, for `LEFT`, `BOTTOM` and `TOP` .
> 
> In addition the images for tests have now been added in `img` folder.

There is no difference. I had made a typo in the doc string but the functionality is correct. This has been fixed in commit https://github.com/nextsimhub/domain_decomp/pull/78/commits/5380006efe2805f9a38084d429203060ca09ffb0

Please remove the image from the img folder. If we want to upload an image it should be svg and have more description. FWIW we can see the domain layout looking at the partition_mask files.

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -32,8 +32,10 @@ create_test(TARGET test_grid_2 SOURCES test_grid_2.cpp)
 create_test(TARGET test_zoltan_0 SOURCES test_zoltan_partitioner_0.cpp)
 create_test(TARGET test_zoltan_1 SOURCES test_zoltan_partitioner_1.cpp)
 create_test(TARGET test_zoltan_2 SOURCES test_zoltan_partitioner_2.cpp)
-create_test(TARGET test_halo_corner_start SOURCES test_halo_corner_start.cpp)
+create_test(TARGET test_haloCornerBufferPositions SOURCES test_haloCornerBufferPositions.cpp)
 create_test(TARGET test_is_corner_neighbour SOURCES test_is_corner_neighbour.cpp)
+create_test(TARGET test_is_neighbour SOURCES test_is_neighbour.cpp)
+create_test(TARGET test_haloBufferPositions SOURCES test_haloBufferPositions.cpp)
```
Comment (joewallwork) [inline test/CMakeLists.txt]: Very minor point: you could consider using a consistent case convention for naming the files. Probably best to make it consistent with nextSIM-DG.

File: Partitioner.hpp [hunk 6]
```diff
@@ -249,50 +243,34 @@ class LIB_EXPORT Partitioner {
...
-     *    │           │     │ │    1     │    │  │          │    │
-     *    │ 0 1 2 3[4]│ 4 5 │ │          │    │  │         4│    │
-     *   2├───────────┤     │ ├──────────┤  2 │  ├──────────┤    │
-     *    │ 5 6 7 8 9 │ 2 3 │ │          │    │  │          │    │
-     *    │           │     │ │    0     │    │  │          │    │
-     * Y▲ │ 0 1 2 3 4 │ 0 1 │ │          │    │  │          │    │
-     *  │ └───────────┴─────┘ └──────────┴────┘  └──────────┴────┘
-     *  │0            5     7
-     *  └───►X
+     * @brief Compute the sendPos and recvPos for halo exchange.
```
Comment (joewallwork) [inline Partitioner.hpp:252]: Suggest spelling this out because it might not be immediately obvious what `sendPos` and `recvPos` mean.

File: Partitioner.cpp [hunk 7]
```diff
@@ -17,6 +17,27 @@
 #include <netcdf.h>
 #include <netcdf_par.h>
 
+/*!
+ * @brief Compute modulo operation with non-negative result
+ *
+ * This function computes a % n and ensures the result is always non-negative.
```
Comment (joewallwork) [inline Partitioner.cpp]: Worth spelling out `a % n` because it might not be obvious. This isn't the same notation that would be used in maths, for example.

File: test/test_haloEdgeBufferPositions.cpp
Comment (joewallwork) [inline test/test_haloEdgeBufferPositions.cpp:1]: Do I recall correctly that we decided to use the term `haloEdgeBuffer` to distinguish from corner case?

File: test/test_haloCornerBufferPositions.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
+/*!
+ * @file halo_corner_start.cpp
```
Comment (joewallwork) [inline test/test_haloCornerBufferPositions.cpp]: Name is inconsistent with filename.

File: test/test_haloBufferPositions.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+#include "Partitioner.hpp"
+#include "Utils.hpp"
+#include <doctest/extensions/doctest_mpi.h>
+
+#include <iostream>
+
+extern int global_argc;
+extern char** global_argv;
+
+MPI_TEST_CASE("Corner neighbour: Non-periodic, 4 MPI ranks", 4)
```
Comment (joewallwork) [inline test/test_haloBufferPositions.cpp]: This isn't really true - is it not edge neighbours? Also, the periodic case is also considered in this test as well as the non-periodic case.

File: test/test_haloCornerBufferPositions.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+#include "Grid.hpp"
+#include "Partitioner.hpp"
+#include "Utils.hpp"
+#include <doctest/extensions/doctest_mpi.h>
+
+extern int global_argc;
+extern char** global_argv;
+
+MPI_TEST_CASE("Corner neighbour: Non-periodic, 4 MPI ranks", 4)
```
Comment (joewallwork) [inline test/test_haloCornerBufferPositions.cpp]: Again, the periodic case is also considered in this test as well as the non-periodic case.

File: test/test_is_neighbour.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
+/*!
+ * @file .cpp
```
Comment (joewallwork) [inline test/test_is_neighbour.cpp]: Missing filename

File: test/test_is_neighbour.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    // Non-periodic cases
+    bool neighbour;
+
+    // RIGHT
+    if (test_rank == 0) {
+        neighbour = partitioner->is_neighbour(domains[0], domains[2], RIGHT);
+        REQUIRE(neighbour);
+    }
```
Comment (joewallwork) [inline test/test_is_neighbour.cpp]: Could we not just do something like this? (And similarly elsewhere.)
```suggestion
    // RIGHT
    if (test_rank == 0) {
        REQUIRE(partitioner->is_neighbour(domains[0], domains[2], RIGHT));
    }
```

**Review by joewallwork — CHANGES_REQUESTED** [review]
Thanks for this. I've not gone through all the working of the subdomains, numbers, neighbours etc because I figured you'd have it right between the two of you. However, I do have some comments and suggestions that it'd be good to address before merging.

File: test/test_haloBufferPositions.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        int startPos;
+        int recvPos;
+    };
+
+    const bool pxOn = true, pyOn = true;
+
+    // The int in tuple<Edge, int> refers to rank. In the case of edge neighbours,
+    // it is possible that one edge may corresspond to multiple neighbour unlike
+    // corner neighbour where a given vertice can correspond to only one neighbour.
+    std::map<std::tuple<Edge, int>, HaloEdgeInfo> edgeInfoExpected, edgeInfoActual;
```
Comment (TomMelt) [inline test/test_haloBufferPositions.cpp]:
```suggestion
    std::map<Edge>, std::list<HaloEdgeInfo>> edgeInfoExpected, edgeInfoActual;
```

The order of this data structure is slightly wrong. We really want a map of Edges with the respective neighbor data as shown above.

It should look something like this:

```
{
 'RIGHT': (haloEdgeInfo1, haloEdgeInfo2, haloEdgeInfo3),
  'LEFT': (haloEdgeInfo1),
   'TOP': (haloEdgeIhaloEdgeInfofo1, haloEdgeInfo2, haloEdgeInfo3, haloEdgeInfo4),
'BOTTOM': (haloEdgeInfo1, haloEdgeInfo2)
}
```

I would also recommend changing from a tuple to a list because you will need to append to it for `edgeInfoActual`

File: test/test_haloEdgeBufferPositions.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    int start;
+    int recv;
+
+    struct HaloEdgeInfo {
+        bool isPeriodic;
+        int startPos;
+        int recvPos;
+    };
```
Comment (TomMelt) [inline test/test_haloEdgeBufferPositions.cpp:71]:
```suggestion
    struct HaloEdgeInfo {
        bool isPeriodic;
        int rank;
        int startPos;
        int recvPos;
    };
```

You need to extend this object `HaloEdgeInfo` to add the rank (as done in the corner test).

File: test/test_haloBufferPositions.cpp [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        edgeInfoExpected[{ TOP, 2 }] = { true, 0, 6 };
+        edgeInfoExpected[{ LEFT, 1 }] = { false, 6, 10 };
+    }
+
+    for (auto edge : edges) {
+        for (int p = 0; p < test_nb_procs; p++) {
+            // periodic neighbours
+            if (partitioner->is_neighbour(domains[test_rank], domains[p], edge, pxOn, pyOn)) {
+                partitioner->haloBufferPositions(domains[test_rank], domains[p], edge, start, recv);
+                edgeInfoActual[{ edge, p }] = { true, start, recv };
```
Comment (TomMelt) [inline test/test_haloBufferPositions.cpp]: You will need to update this logic, to append the neighbor info one at a time.

Once you have changed `edgeInfoExpected` and `edgeInfoActual` to lists, you will be able to append.

It may look something like, you would need to check it exists before you try to append for example. 

```suggestion
                edgeInfoActual.at(edge).append(true, p, start, recv);
```

First check it exists, if yes then append. If no, then create a list with 1 item and store it in the map for that specific edge.

File: Partitioner.cpp (hunk 7 above)
Comment (TomMelt) [reply Partitioner.cpp]: Thanks @joewallwork done in e23165f

File: test/CMakeLists.txt (hunk 5 above)
Comment (TomMelt) [reply test/CMakeLists.txt]: Thanks.... This is on my todo list already :sweat: I made issue #80

I will make it a separate PR to address though.

File: test/test_haloEdgeBufferPositions.cpp
Comment (TomMelt) [reply test/test_haloEdgeBufferPositions.cpp:1]: yes this will also be done as part of a separate PR fixing #80

File: test/test_is_neighbour.cpp (hunk 11 above)
Comment (TomMelt) [reply test/test_is_neighbour.cpp]: thanks done in 8b98a58

File: test/test_haloCornerBufferPositions.cpp (hunk 8 above)
Comment (TomMelt) [reply test/test_haloCornerBufferPositions.cpp]: thanks done in 8b98a58

File: Partitioner.hpp (hunk 6 above)
Comment (TomMelt) [reply Partitioner.hpp:252]: Thanks @joewallwork I think it got lost in the diff. But I had added description to the vars below:

https://github.com/nextsimhub/domain_decomp/blob/8b98a5870b2c7d895897418c546afbdadad0baeb/Partitioner.hpp#L252-L261

Do you want more explanation or is it ok as is?

File: test/test_is_neighbour.cpp (hunk 12 above)
Comment (TomMelt) [reply test/test_is_neighbour.cpp]: I just realised that the testing of `partitioner->is_neighbour` is now much better handled in `test_haloBufferPositions.cpp`. So I would prefer to remove this file altogether.

Sorry for the additional review load :grimacing: 

My problem with this specific file is it doesn't check all cases. Just one for each rank. Whereas the `test_haloBufferPositions.cpp` is designed to check all of them.

File: test/test_haloBufferPositions.cpp (hunk 9 above)
Comment (TomMelt) [reply test/test_haloBufferPositions.cpp]: thanks done in a81b85c

File: test/test_haloCornerBufferPositions.cpp (hunk 10 above)
Comment (TomMelt) [reply test/test_haloCornerBufferPositions.cpp]: thanks done in a81b85c

File: Partitioner.hpp (hunk 6 above)
Comment (joewallwork) [reply Partitioner.hpp:252]: Ah okay got it, thanks.

File: test/test_is_neighbour.cpp (hunk 12 above)
Comment (joewallwork) [reply test/test_is_neighbour.cpp]: Sure, no worries. Will re-review now.

File: Partitioner.cpp [hunk 16]
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
...
-    int start = 0;
-    if (vertex == TOP_LEFT) {
-        if (d2.p1.y < top) {
-            start = (top - d2.p1.y + 1) * d2.get_width() - 1;
+    send_pos = 0;
+    if (vertex == TOP_RIGHT) {
+        // for a TOP_RIGHT vertex the corner neighbour can either be along the other domains bottom
+        // or left edge. We need to check so we know where to look in the send buffer.
+        xpos = mod((d1.p2.x), globalX);
+        ypos = mod((d1.p2.y), globalY);
```
Comment (joewallwork) [inline Partitioner.cpp]:
```suggestion
        xpos = mod(d1.p2.x, globalX);
        ypos = mod(d1.p2.y, globalY);
```

File: Partitioner.cpp [hunk 17]
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
...
+            // corner of both e.g., the bottom left corner)
+            int dx = xpos - d2.p1.x;
+            send_pos = dx;
         }
-    } else if (vertex == TOP_RIGHT) {
-        if (d2.p1.x < right) {
-            start = right - d2.p1.x;
+    } else if (vertex == TOP_LEFT) {
+        xpos = mod((d1.p1.x - 1), globalX);
+        ypos = mod((d1.p2.y), globalY);
```
Comment (joewallwork) [inline Partitioner.cpp]:
```suggestion
        xpos = mod(d1.p1.x - 1, globalX);
        ypos = mod(d1.p2.y, globalY);
```

File: Partitioner.cpp [hunk 18]
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
...
-            start = (top - d2.p1.y) * d2.get_width();
+            int dx = xpos - d2.p1.x;
+            send_pos = dx;
         }
-    } else if (vertex == BOTTOM_RIGHT) {
-        if (d2.p1.y < bottom) {
-            start = (bottom - d2.p1.y - 1) * d2.get_width() + (right - d2.p1.x);
+    } else if (vertex == BOTTOM_LEFT) {
+        xpos = mod((d1.p1.x - 1), globalX);
+        ypos = mod((d1.p1.y - 1), globalY);
```
Comment (joewallwork) [inline Partitioner.cpp]:
```suggestion
        xpos = mod(d1.p1.x - 1, globalX);
        ypos = mod(d1.p1.y - 1, globalY);
```

File: Partitioner.cpp [hunk 19]
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
...
-            start = (bottom - d2.p1.y - 1) * d2.get_width();
+            int dx = xpos - d2.p1.x;
+            send_pos = d2.get_width() + d2.get_height() + dx;
         }
-    } else if (vertex == BOTTOM_LEFT) {
-        if (d2.p1.y < bottom) {
-            start = (bottom - d2.p1.y - 1) * d2.get_width() + (left - d2.p1.x) - 1;
+    } else if (vertex == BOTTOM_RIGHT) {
+        xpos = mod((d1.p2.x), globalX);
+        ypos = mod((d1.p1.y - 1), globalY);
```
Comment (joewallwork) [inline Partitioner.cpp]:
```suggestion
        xpos = mod(d1.p2.x, globalX);
        ypos = mod(d1.p1.y - 1, globalY);
```

**Review by joewallwork — APPROVED** [review]
I just have some minor formatting suggestions but otherwise I think it looks good.

File: Partitioner.cpp (hunk 19 above)
Comment (TomMelt) [reply Partitioner.cpp]: Thanks @joewallwork for the suggestions. I have applied them all 👌
